keps/sig-cli/kustomize-annotate-generated.md

```diff
@@ -0,0 +1,119 @@
+---
+title: Kustomize Annotate Generated Secrets and ConfigMaps
+authors:
+ - "@pwittrock"
+owning-sig: sig-cli
+participating-sigs:
+ - sig-cli
+reviewers:
+ - "@janetkuo"
+ - "@anguslees"
+approvers:
+ - "@monopole"
+editors:
+ - "@pwittrock"
+creation-date: 2019-03-01
+last-updated: 2019-03-01
+status: provisional
+see-also:
+replaces:
+superseded-by:
+ - n/a
+---
+
+# Kustomize Annotate Generated Secrets and ConfigMaps
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+- Annotate generated Secrets and ConfigMaps with Workloads the reference them
+- Annotated Workloads with the Secrets and ConfigMaps they reference
+- Workload annotations are retained after the Workloads stop referencing the Secrets and ConfigMaps
+- Secret and ConfigMap annotations are retained after the Workloads stop referencing them
+
+## Motivation
+
+- Kustomize generates Secret and ConfigMap Resources which are referenced by Workloads
+  - When the data for the Secrets or ConfigMaps change, new Secret or ConfigMap Resources are generated
+  - Kustomize updates the Workloads to refer to the new Resources, and stop referring to the old Resources
+  - References to the old Secret and ConfigMap Resources are leaked when new Resources are generated
+  - The information that the Resources were generated by Kustomize rather than explicitly defined by
+    users is lost
+
+### Goals
+
+- Provide provenance for Resources generated by Kustomize
+  - Retain a record of Secret and ConfigMap Resources generated by Kustomize.
+  - Retain a record of which Workloads each generated Secret and ConfigMap was consumed by.
+  - Retain a record of which generated Secret and ConfigMap Resources each Workload consumed.
+- Allow other tools to identify generated Secrets and ConfigMaps that are no longer in use as
+  candidates for garbage collection.
+
+### Non-Goals
+
+## Proposal
+
+Kustomize already modifies Workloads that refer to generated Secrets and ConfigMaps by updating
+the names to match the generated names.
+
+### Annotations on Workloads
+
+Set an annotation on each Workload for each generated Secret and ConfigMap it references.  When
```

Comment:
Should these be on the Deployment/Statefulset/Job/etc or on the Pod (via the ReplicaSet)?  (eg: in Deployment `spec.template.metadata.annotations`)

The advantage of storing it in the Pod is that we can theoretically track references from multiple versions of the Deployment, and ensure the configmap/secret still exists for as long as the Deployment is able to rolled back (ie: as long as the ReplicaSet exists).

Reply:
With the proposed design, the Deployment will have annotations for all of the ConfigMaps it has ever referenced.  The idea being when the Deployment stops referring to it, some body could write a tool to GC the unused ConfigMaps.  Without this there isn't a record of past ConfigMaps that kustomize has generated.

Reply:
wait what?  We're going to keep _all_ the configmaps that were _ever_ referenced, even though this long-lived Deployment stopped referencing that particular (hashed) configmap 3 years ago?

Even more so: please include a description in this doc of how we intend to actually use these annotations, because that seems ... bonkers.

Reply:
there is no gc for the configmaps themselves...so they will be around forever as well :(.  writing these annotations allows a tool to be developed which could find and gc unused configmaps, and then remove the annotations.